Common/Utility/ExternalModCallUtils.cs

```diff
@@ -1,54 +1,110 @@
-﻿using Terraria;
+﻿using CalamityMod.Projectiles.Magic;
+using Mono.Cecil.Cil;
+using System.Collections.Generic;
+using Terraria;
 using Terraria.ModLoader;
 
 namespace TheBereftSouls.Common.Utility
 {
     public class ExternalModCallUtils
     {
-        public static int GetItemFromMod(Mod mod, string item)
+        public struct CacheItem
         {
-            return mod.Find<ModItem>(item).Type;
+            public Mod mod;
+            public string name;
+            public ModItem instance;
         }
+        public struct CacheProjectile
+        {
+            public Mod mod;
+            public string name;
+            public ModProjectile instance;
+        }
+        public struct CacheNpc
+        {
+            public Mod mod;
+            public string name;
+            public ModNPC instance;
+        }
+        public static HashSet<CacheItem> ItemsCache { get; set; } = [];
+        public static ModItem GetItemFromMod(Mod mod, string item)
+        {
+            foreach (CacheItem cache in ItemsCache)
+            {
+                if (cache.mod == mod && cache.name == item)
+                    return cache.instance;
+            }      
```

Comment:
There's no point of "caching" if we'll iterate through everything. That's an O(n) operation every time. Microsoft has made caching extensions to simplify the process<sup>\[[1]\]</sup>. Here's an example implementation they have with time based caching:

- https://learn.microsoft.com/en-us/dotnet/core/extensions/caching#put-it-all-together

There's also a great geeksforgeeks article tutorial for LRU cache<sup>\[[2]\]</sup> as well as examples in C#, Java, Python, C++ and JavaScript<sup>\[[3]\]</sup> of an LRU cache implementation.

> [!NOTE]
> I recommend implementing LRU caching because of the more control over how much memory there is. If you're up for the challenge you can implement LRU caching + timeouts for the cache but that's more complicated and usually the GC cleans up any redundant memory.

[1]: https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory?view=net-8.0 "Microsoft.Extensions.Caching.Memory Namespace"

[2]: https://www.geeksforgeeks.org/lru-cache-implementation/ "LRU cache implementation tutorial"

[3]: https://www.geeksforgeeks.org/lru-cache-implementation-using-double-linked-lists/ "LRU cache implementation examples"